File: OI/jzoj/1737.pas
const maxn=101000;
type node=record
       son,ans:longint;
     end;
     arr=array[0..4]of longint;
var next,y,cost:array[0..maxn*2]of longint;
    g:array[0..maxn]of longint;
    f,fl,fr:array[0..maxn]of node;
    n,i,j,u,v,co,tt,answer:longint;
function max(a,b:longint):longint;
begin
    if a>b then exit(a);exit(b);
end;
procedure insert(i,j,v:longint);
begin
    inc(tt);
    next[tt]:=g[i];
    g[i]:=tt;
    y[tt]:=j;
    cost[tt]:=v;
end;
procedure dfs(r,fa:longint);
var  i,p:longint;
begin
    f[r].son:=0;
    f[r].ans:=0;
    p:=g[r];
    while p<>0 do begin
      i:=y[p];
      if i<>fa then begin
         dfs(i,r);
         f[r].ans:=max(f[r].ans,f[r].son+f[i].son+cost[p]);
         f[r].son:=max(f[r].son,f[i].son+cost[p]);
      end;
      p:=next[p];
    end;
end;
procedure ready(p,fa,f1,f2:longint;var bf1,bf2:longint);
var i,jf1,jf2:longint;
begin
    if p=0 then begin
       bf1:=0;bf2:=0;
       exit;
    end;
    i:=y[p];
    if i=fa then begin
       ready(next[p],fa,f1,f2,bf1,bf2);
       exit;
    end;
    fl[i].son:=f1;
    fl[i].ans:=f2;
    ready(next[p],fa,max(f1,f[i].son+cost[p]),max(f2,max(f[i].ans,f1+f[i].son+cost[p])),jf1,jf2);
    fr[i].son:=jf1;fr[i].ans:=jf2;
    bf1:=max(jf1,f[i].son+cost[p]);
    bf2:=max(jf2,max(f[i].ans,jf1+f[i].son+cost[p]));
end;
procedure dp(r,fa,f1,f2:longint);
var p,i,jf1,jf2,tmp:longint;
begin
    ready(g[r],fa,f1,f2,jf1,jf2);
    p:=g[r];
    while p<>0 do begin
      i:=y[p];
      if i<>fa then begin
         jf1:=max(fl[i].son,fr[i].son)+cost[p];
         jf2:=max(fl[i].son+fr[i].son,max(fl[i].ans,fr[i].ans));
	 tmp:=jf2+f[i].ans;
         answer:=max(answer,tmp);
         dp(i,r,jf1,jf2);
      end;
      p:=next[p];
    end;
end;
begin
	assign(input,'1737.in');reset(input);
    readln(n);
    tt:=0;
    for i:=1 to n-1 do begin
        readln(u,v,co);
        insert(u,v,co);
        insert(v,u,co);
    end;
    dfs(1,-1);
    dp(1,-1,0,0);
    writeln(answer);
end.

File: OI/jzoj/1737.cpp
#include <cstdio>
#include <algorithm>
#include <cstring>
#include <cmath>

using namespace std;

const int N=100005,M=200005;
int n,tot,ans;
int g[N],h[N];
int v[M],w[M],next[M];
int f[N][2][4],id[N][2][3];

void add(int x,int y,int z)
{
	v[++tot]=y;
	w[tot]=z;
	next[tot]=g[x];
	g[x]=tot;
}

void dfs(int x,int y)
{
	for(int i=g[x];i;i=next[i])
	{
		if (v[i]==y) continue;
		dfs(v[i],x);
		h[x]=max(h[x],h[v[i]]);
		if (h[v[i]]>f[x][0][1])
		{
			f[x][0][2]=f[x][0][1];
			f[x][0][1]=h[v[i]];
			id[x][0][1]=v[i];
		} 
		else if (h[v[i]]>f[x][0][2]) f[x][0][2]=h[v[i]];
		int t=f[v[i]][1][1]+w[i];
		if (t>f[x][1][1])
		{
			f[x][1][3]=f[x][1][2];
			f[x][1][2]=f[x][1][1];
			f[x][1][1]=t;
			id[x][1][2]=id[x][1][1];
			id[x][1][1]=v[i];
		}
		else if (t>f[x][1][2])
		{
			f[x][1][3]=f[x][1][2];
			f[x][1][2]=t;
			id[x][1][2]=v[i];
		}
		else if (t>f[x][1][3]) f[x][1][3]=t;
	}
	h[x]=max(f[x][1][1]+f[x][1][2],f[x][0][1]);
}

void work(int x,int fa,int y,int z)
{
	int t1,t2,t3;
	if (x!=1) ans=max(ans,h[x]+y);
	for(int i=g[x]; i; i=next[i])
	{
		if  (v[i]==fa) continue;
		if  (v[i]==id[x][1][1])
		{
			t1 = f[x][1][2];
			t2 = f[x][1][2] + f[x][1][3];
		} else
		{
			t1 =f[x][1][1];
			if (v[i]==id[x][1][2]) t2 = f[x][1][1] + f[x][1][3];
			else t2 = f[x][1][1] + f[x][1][2];
		}
		if (v[i]==id[x][0][1]) t3 = f[x][0][2]; else t3 = f[x][0][1];
		work(v[i],x,max(max(max(y,t3),t1+z),t2),max(z,t1)+w[i]);
	}
}

int main()
{
	freopen("1737.in","r",stdin);
	scanf("%d\n",&n);
	
	for(int i=1; i<n; i++)
	{
		int x,y,z;
		scanf("%d%d%d",&x,&y,&z);
		add(x,y,z);
		add(y,x,z);
	}
	dfs(1,0);
	work(1,0,0,0);
	
	printf("%d\n",ans);
	return 0;
}

